syntax = "proto3";

import "google/protobuf/any.proto";

option go_package = "github.com/xiaonanln/goverse/proto/goverse_pb";


package goverse;


service Goverse {
  rpc Status (Empty) returns (StatusResponse);
  rpc ListObjects (Empty) returns (ListObjectsResponse);
  rpc CallObject (CallObjectRequest) returns (CallObjectResponse);
  rpc CreateObject (CreateObjectRequest) returns (CreateObjectResponse);
  rpc PushMessageToClient (PushMessageToClientRequest) returns (PushMessageToClientResponse);
}

message StatusResponse {
  string advertise_addr = 1;
  int64 num_objects = 2;
  int64 uptime_seconds = 3;
}

message CallObjectRequest {
  string id = 1;
  string method = 2;
  google.protobuf.Any request = 4;
}

message CallObjectResponse {
  google.protobuf.Any response = 1;
}

message CreateObjectRequest {
  string type = 1;
  google.protobuf.Any init_data = 2;
  string id = 3; // Optional: if not provided, the node will generate an ID
}

message CreateObjectResponse {
  string id = 1;
}

message ObjectInfo {
  string type = 1;
  string id = 2;
  int64 creation_time = 3; // Unix timestamp in seconds
}

message ListObjectsResponse {
  repeated ObjectInfo objects = 1;
}

message PushMessageToClientRequest {
  string client_id = 1;
  google.protobuf.Any message = 2;
}

message PushMessageToClientResponse {
  // Empty response - success is indicated by no error
}

message Empty {}
